File: public/ocr-cleanup.js
/*
    Copyright (C) 2025, Paul Hammant

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as
    published by the Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

/**
 * OCR-specific artifact cleanup functions
 * Only needed for camera/OCR verification paths, not for text selection
 */

/**
 * Clean OCR artifacts from text (border characters, trailing letters)
 * Call this BEFORE normalizeText() for OCR'd content
 * @param {string} text - Raw OCR text
 * @returns {string} Text with OCR artifacts removed
 */
function cleanOcrArtifacts(text) {
    const lines = text.split('\n');

    const cleanedLines = lines.map(line => {
        // Remove leading border artifacts (OCR from registration marks/borders)
        // Common characters: | ~ ` ^ * # + = _ \ / [ ] { }
        // Examples: "| text", "~ text", "|| text", "| | text"
        line = line.replace(/^[|~`^*#+=/\\_\[\]{}]+\s*/, '');

        // Remove trailing border artifacts (OCR from registration marks/borders)
        // Examples: "text |", "text ~", "text ||", "text | |"
        line = line.replace(/\s*[|~`^*#+=/\\_\[\]{}]+$/, '');

        // Remove trailing single lowercase letter (OCR artifact)
        // Examples: "John Doe a", "Company Name b" -> "John Doe", "Company Name"
        // Preserves uppercase like "Appendix A", "Grade A" which are likely meaningful
        line = line.replace(/\s+[a-z]$/, '');

        return line;
    });

    return cleanedLines.join('\n');
}

// Export for Node.js testing (doesn't affect browser usage)
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { cleanOcrArtifacts };
}

File: apps/android/app/src/main/java/com/liveverify/app/OcrCleanup.kt
/*
    Copyright (C) 2025, Paul Hammant

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as
    published by the Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

package com.liveverify.app

/**
 * OCR-specific artifact cleanup functions.
 * Matches public/ocr-cleanup.js - keep in sync.
 */
object OcrCleanup {

    /**
     * Clean OCR artifacts from text (border characters, trailing letters).
     * Call this BEFORE normalizeText() for OCR'd content.
     *
     * @param text Raw OCR text
     * @return Text with OCR artifacts removed
     */
    fun cleanOcrArtifacts(text: String): String {
        val lines = text.split("\n")

        val cleanedLines = lines.map { line ->
            var cleaned = line

            // Remove leading border artifacts (OCR from registration marks/borders)
            // Common characters: | ~ ` ^ * # + = _ \ / [ ] { }
            cleaned = cleaned.replace(Regex("^[|~`^*#+=/\\\\\\[\\]{}]+\\s*"), "")

            // Remove trailing border artifacts
            cleaned = cleaned.replace(Regex("\\s*[|~`^*#+=/\\\\\\[\\]{}]+$"), "")

            // Remove trailing single lowercase letter (OCR artifact)
            // Preserves uppercase like "Appendix A", "Grade A" which are likely meaningful
            cleaned = cleaned.replace(Regex("\\s+[a-z]$"), "")

            cleaned
        }

        return cleanedLines.joinToString("\n")
    }
}
